spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/integration/TraceContextPropagationChannelInterceptor.java

```diff
@@ -0,0 +1,203 @@
+import static org.springframework.cloud.sleuth.TraceContextHolder.getCurrentSpan;
+import static org.springframework.cloud.sleuth.TraceContextHolder.setCurrentSpan;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.springframework.aop.support.AopUtils;
+import org.springframework.cloud.sleuth.Span;
+import org.springframework.cloud.sleuth.event.ClientReceivedEvent;
+import org.springframework.cloud.sleuth.event.ClientSentEvent;
+import org.springframework.context.ApplicationEvent;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.context.ApplicationEventPublisherAware;
+import org.springframework.integration.channel.DirectChannel;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.ChannelInterceptorAdapter;
+import org.springframework.messaging.support.ExecutorChannelInterceptor;
+import org.springframework.util.Assert;
+
+/**
+ * The {@link ExecutorChannelInterceptor} implementation responsible for
+ * the {@link Span} propagation from one message flow's thread to another
+ * through the {@link MessageChannel}s involved in the flow.
+ * <p>
+ * In addition this interceptor cleans up (restores) the {@link Span}
+ * in the containers Threads for channels like
+ * {@link org.springframework.integration.channel.ExecutorChannel}
+ * and {@link org.springframework.integration.channel.QueueChannel}.
+ * @author Spencer Gibb
+ * @since 1.0
+ */
+public class TraceContextPropagationChannelInterceptor
+		extends ChannelInterceptorAdapter implements ExecutorChannelInterceptor, ApplicationEventPublisherAware {
+
+	private final static ThreadLocal<Span> ORIGINAL_CONTEXT = new ThreadLocal<>();
+
+	private ApplicationEventPublisher publisher;
+
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher publisher) {
+		this.publisher = publisher;
+	}
+
+	@Override
+	public final Message<?> preSend(Message<?> message, MessageChannel channel) {
+		if (DirectChannel.class.isAssignableFrom(AopUtils.getTargetClass(channel))) {
+			return message;
+		}
+
+		//TODO: start span from headers?
+		Span span = getCurrentSpan();
+
+		if (span != null) {
+			publish(new ClientSentEvent(this, span));
+			return new MessageWithSpan(message, span);
+		}
+		else {
+			return message;
+		}
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public final Message<?> postReceive(Message<?> message, MessageChannel channel) {
+		if (message instanceof MessageWithSpan) {
+			MessageWithSpan messageWithSpan = (MessageWithSpan) message;
+			Message<?> messageToHandle = messageWithSpan.message;
+			populatePropagatedContext(messageWithSpan.span, messageToHandle, channel);
+
+			publish(new ClientReceivedEvent(this, messageWithSpan.span));
+			return message;
```

Comment:
I'm returning the `message` itself and not `messageWithSpan.message` because I want the combined headers in `message`.  Is this safe? Another alternative would be to return `messageWithSpan.message` wrapped with `message`'s headers.